### PR TITLE
expression: Support GBK for builtin function AesEncrypt.

### LINKED ISSUE #29168: Support GBK for builtin function `AesEncrypt`

### PR DESCRIPTION
**Pull request by jackwener**
### What problem does this PR solve?

Issue Number: close #29168

### What is changed

* Add charset convert for AesEncrypt args
* Add explain tests

### Check List

Tests

* [x] Unit test
* [x] Integration test

Side effects

* [x] None

Documentation

* [x] None

### Release note

```
 None
```

### CODE REVIEW

**Comment by jackwener** [comment]
It's only remaining issue in #28876 . Due to my problem, the problem was not completed quickly.

**Comment by unconsolable** [comment]
Could you please take a look at https://github.com/pingcap/tidb/pull/29736 and try on it? IMHO all the arguments of `aes_encrypt` can be wrapped by function `to_binary` and reduce code repetition.

**Comment by tangenta** [comment]
I think we don't have to change the code after https://github.com/pingcap/tidb/pull/29950 is merged. @jackwener Could you please update this PR and only keep the tests?

**Comment by jackwener** [comment]
> I think we don't have to change the code after #29950 is merged. @jackwener Could you please update this PR and only keep the tests?

OK

**Comment by jackwener** [comment]
@zimulala @tangenta PTAL 🤞

**Comment by unconsolable** [comment]
https://github.com/pingcap/tidb/blob/e45387309b2f05c52b89f6ecd6566ad79d5ba6a4/expression/builtin_convert_charset.go#L111-L116
Are you sure the test is right? #29950 don't add `ast.AesEncrypt` to `toBinaryMap`

**Comment by unconsolable** [comment]
https://github.com/pingcap/tidb/blob/e45387309b2f05c52b89f6ecd6566ad79d5ba6a4/expression/builtin_convert_charset.go#L111-L116
But #29950 don't add `ast.AesEncrypt` to `toBinaryMap`, neither this PR does. Does it actually implemented?

**Comment by jackwener** [comment]
> https://github.com/pingcap/tidb/blob/e45387309b2f05c52b89f6ecd6566ad79d5ba6a4/expression/builtin_convert_charset.go#L111-L116
> 
> 
> But #29950 don't add `ast.AesEncrypt` to `toBinaryMap`, neither this PR does. Does it actually implemented?

In test, In found that it don't need add `Encrypt` and `Decrypt` in toBinaryMap.

**Comment by jackwener** [comment]
> https://github.com/pingcap/tidb/blob/e45387309b2f05c52b89f6ecd6566ad79d5ba6a4/expression/builtin_convert_charset.go#L111-L116
> 
> 
> But #29950 don't add `ast.AesEncrypt` to `toBinaryMap`, neither this PR does. Does it actually implemented?

Added

**Comment by jackwener** [comment]
Ok, the result is same with `MySQL`

File: expression/builtin_encryption_test.go [hunk 1]
```diff
@@ -164,6 +164,47 @@ func TestAESEncrypt(t *testing.T) {
...
+		{"aes-128-ecb", "gbk", gbkStr, []interface{}{"123"}, "6AFA9D7BA2C1AED1603E804F75BB0127"},
+		{"aes-128-ecb", "utf8mb4", "123", []interface{}{"你好"}, "E03F6D9C1C86B82F5620EE0AA9BD2F6A"},
+		{"aes-128-ecb", "gbk", "123", []interface{}{gbkStr}, "31A2D26529F0E6A38D406379ABD26FA5"},
+		{"aes-128-ecb", "utf8mb4", "你好", []interface{}{"你好"}, "3E2D8211DAE17143F22C2C5969A35263"},
+		{"aes-128-ecb", "gbk", gbkStr, []interface{}{gbkStr}, "84982910338160D037615D283AD413DE"},
+		// test for cbc
+		{"aes-128-cbc", "utf8mb4", "你好", []interface{}{"123", "1234567890123456"}, "B95509A516ACED59C3DF4EC41C538D83"},
+		{"aes-128-cbc", "gbk", gbkStr, []interface{}{"123", "1234567890123456"}, "D4322D091B5DDE0DEB35B1749DA2483C"},
+		{"aes-128-cbc", "utf8mb4", "123", []interface{}{"你好", "1234567890123456"}, "E19E86A9E78E523267AFF36261AD117D"},
+		{"aes-128-cbc", "gbk", "123", []interface{}{gbkStr, "1234567890123456"}, "5A2F8F2C1841CC4E1D1640F1EA2A1A23"},
```
Comment (unconsolable) [inline expression/builtin_encryption_test.go:188]: Please take a look of https://github.com/pingcap/tidb/pull/30072, we'd better leave the conversion to `to_binary`.

Comment (zimulala) [reply expression/builtin_encryption_test.go:188]: Please refer to @unconsolable's PR to modify the usage of `datumsToConstants`.

**Comment by zimulala** [comment]
@jackwener Why did you close the PR?

**Review by tangenta — APPROVED** [review]

**Review by Defined2014 — APPROVED** [review]

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**